### ISSUE TITLE
Ping is not working

### ISSUE THREAD

**Issue by patryk9200** [Microsoft/BashOnWindows#18]
Ping is not working on WSL.

When used, returns:
**# ping google.com
ping: icmp open socket: Socket type not supported**

**Comment by fhelje** [comment]
Have the same issue network doesn't work, running windows in parallels on a mac

**Comment by gdude2002** [comment]
I got networking working on mine by disabling my AV's firewall, but it doesn't fix `ping`.

**Comment by patryk9200** [comment]
For me network is working, I can sucessfully download content from internet for example: wget google.com but what is not working is ping command and many other network related stuff.

**Comment by jolibert** [comment]
I confirm that requests out of a normal, _SYN/SYN-ACK/ACK_ TCP connection (transport -_UDP_- and network -_ICMP even locally_- layers on ISO model) are not working.
Normally it is looking good to disallow communications at the network level and below, but for internal developments, it should be considered as back-end business may be related.

**Comment by kjk** [comment]
Same issue when using 14316 in latest Parallels on latest Mac OS X.

Inside bash, network doesn't seem to work. DNS resolution fails, ping a fixed address doesn't work:
```
root@localhost:~# ping google.com
ping: unknown host google.com
root@localhost:~# ping 8.8.8.8
ping: icmp open socket: Socket type not supported
````

The same commands work inside cmd.exe.

This means that `apt-get` doesn't work so I can't install any additional packages.

I tried both "Shared Network" and "Bridged Network" networking modes of Parallels.

It might be deficiency of (or incompatibility with) Parallel's network driver.

Happy to provide more info from within bash (given that network stack is probably emulated in WSK I wouldn't know what behavior standard linux network debugging techniques is buggy but expected vs. unexpectedly buggy).

**Comment by nanogennari** [comment]
@kjk It loos like there are two different issues, one is with DNS resolutions, and it is easily solved, just add a DNS server to /etc/resov.conf, it shoud look like:

```
root@localhost:/# cat /etc/resolv.conf
nameserver 10.10.10.1
```

After that `apt-get` shoud work.

The second issue is with ping, i'm having the same problem here:

```
root@localhost:/# cat /etc/resolv.conf
nameserver 10.10.10.1
```

**Comment by eternelpanic** [comment]
Traceroute is no better:
root@localhost:~# traceroute www.internode.on.net
traceroute to www.internode.on.net (150.101.140.197), 30 hops max, 60 byte packets
setsockopt IP_MTU_DISCOVER: Invalid argument

**Comment by SvenGroot** [comment]
ICMP is currently not supported, which is why "ping" fails. We realize that this makes it inconvenient to diagnose possible networking issues, and are working to address this.

**Comment by patryk9200** [comment]
@SvenGroot  Thanks for info!

**Comment by arjennienhuis** [comment]
Why is this closed?

**Comment by gdude2002** [comment]
..huh, I didn't get an email for this.

Is it fixed? I still get the same behaviour.

**Comment by sunilmut** [comment]
Yes, any network tools that rely on NETLINK or ICMP do not currently work and we are aware of this. We will be tracking\prioritizing which features to light up next by looking at the feedback provided at the wpdev portal. We appreciate your feedback and please make sure to provide it through the wpdev portal as well. Looks like there is already a post tracking this which you can up vote (or add a new one):

https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13383789-enable-network-interface-enumeraration

**Comment by arjennienhuis** [comment]
Closing the issue is still a bad idea and also a bit rude.

A nicer way is to tag it as 'wishlist', 'feature' or even 'backlog'

**Comment by fungus** [comment]
Not only is closing the issue a bad idea, referencing a feedback post that has a description completely unrelated to this issue is really bad.  

No one viewing that feedback post will think it relates to this issue.

**Comment by sunilmut** [comment]
@arjennienhuis, @fungus - Feedback taken regarding closing the issue. I will keep this in mind going forward.

**Comment by defconoi** [comment]
I have multiple networking tools that I cannot use due to this bug effecting ping, dig and anything raw socket related.  Please address and prioritize this bug since Linux networking is a core feature of Linux and this bug probably spans thousands of Linux packages.

**Comment by williambargent** [comment]
My bash can connect to the internet fine and DNS works however I get the same problem with ping and traceroute. 
Traceroute - 'setsockopt IP_MTU_DISCOVER: Invalid argument'
Ping - 'ping: icmp open socket: Socket type not supported'

**Comment by sunilmut** [comment]
@williambargent - ping has been enabled for admin prompts, but hasn't made it to the release branch yet. As for the other network utilities, that's something that we are aware of, but haven't had a chance to enable them yet. You can help us prioritize this work for the future by providing your feedback through our [User Voice page](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo) or up voting this [relevant request](https://wpdev.uservoice.com/forums/266908-command-prompt-console-bash-on-ubuntu-on-windo/suggestions/13383789-enable-network-interface-enumeraration). And, thanks for trying out WSL!

**Comment by russalex** [comment]
Ping works if you run Bash in an admin console as of [14352](https://blogs.windows.com/windowsexperience/2016/05/26/announcing-windows-10-insider-preview-build-14352/).  It is not ideal but is the best we will be able to get it for the Anniversary Edition.  If you try running in a non Admin console you should get a permission denied message.

As @sunilmut points out, please keep adding / up voting on the User Voice page.  We really do look at those when determining priorities.

**Comment by sunilmut** [comment]
Regarding the requirement for admin for ping, that is something we (WSL) have inherited from the Windows TCP/IP stack which requires admin privileges to open RAW sockets for security measures. See [this](https://msdn.microsoft.com/en-us/library/windows/desktop/ms740548(v=vs.85).aspx) MSDN post for more details.

**Comment by paladox** [comment]
ping google.com works for me now.

**Comment by patryk9200** [comment]
Ping is working, thanks!
I will close the issue now.

**Comment by clarity99** [comment]
not working on anniversary update
 root@ROBDOMA:~# ping 192.168.1.144
ping: icmp open socket: Permission denied

**Comment by stuwil** [comment]
@clarity99 You need to open Bash with administrative privileges for it to work.

**Comment by muralimohanputta** [comment]
even in Admim mode its not working  for me

**Comment by sunilmut** [comment]
@muralimohanputta - Can you post the error that you are seeing?

**Comment by muralimohanputta** [comment]
ping: icmp open socket: Permission denied

On Tue, Aug 30, 2016 at 11:16 AM, Sunil Muthuswamy <notifications@github.com

**Comment by sunilmut** [comment]
@muralimohanputta - That error should only show up if the bash console window is non-admin. I know you have indicated earlier that this error is also showing up in admin mode, but I just wanted to clarify that admin mode is not the same as sudo. Can you confirm\double check that you are launching the bash console window as a Windows Administrator?

**Comment by muralimohanputta** [comment]
Hi Sunil,

Thank you. now ping is working for me. I didn't launch the bash shell as a
Windows Administrator.

ifconfig is not working

Warning: cannot open /proc/net/dev (No such file or directory). Limited
output.

I am able to use apt-get install.



On Tue, Aug 30, 2016 at 11:06 PM, Sunil Muthuswamy <notifications@github.com

**Comment by rhavens** [comment]
It would be great if you added support for traceroute, ICMP still seems to be unsupported in this case

**Comment by sunilmut** [comment]
@rhavens - Thanks for your feedback. Yes, we are aware of the lack of support for 'traceroute' and something we are looking at. It's also covered in many other GitHub issues, such as #69.

**Comment by buggsdummy** [comment]
This needs to be re-opened! Thanks!

**Comment by aseering** [comment]
@buggsdummy -- thanks for your post, but if you would like this to be re-opened, please provide enough information about why it needs to be re-opened that someone could potentially fix it.  For example, the specific symptoms that you are seeing.  Please also confirm that you are still seeing the problem on the latest Windows Insider build (currently OS Build 14965).

**Comment by amrith** [comment]
Still does not work. 

$ ping 192.168.115.10
ping: icmp open socket: Permission denied

ssh and telnet appear to work fine. Just ping. I guess the issue is limited to ICMP.

**Comment by aseering** [comment]
@amrith -- is this fixed if you upgrade to a more-recent Insider build?

**Comment by amrith** [comment]
Nope. on the latest as on 2h ago ;)

-amrith


-------- Original message --------

**Comment by sunilmut** [comment]
@amrith - The permission error from ping will only show up if you are running the anniversary update build, and launching bash as Windows non-admin. Admin is no longer a requirement for ping for the Windows Insider build. 
See the previous thread on this post:
>That error should only show up if the bash console window is non-admin. I know you have indicated earlier that this error is also showing up in admin mode, but I just wanted to clarify that admin mode is not the same as sudo. Can you confirm\double check that you are launching the bash console window as a Windows Administrator?

**Comment by amrith** [comment]
I am absolutely *NOT* running bash as a windows administrator. Bash is being run by a standard user and ping is also being run by a non-root user.

 

Is it now the case that bash is only for a Windows Administrator user?

 

-amrith

**Comment by aseering** [comment]
@amrith -- please confirm your exact Windoes build number?

**Comment by amrith** [comment]
I am running Windows 10 Enterprise, Version 1607, OS Build 14393.447.

**Comment by aseering** [comment]
@amrith -- that is not the latest Insider build.  The latest Insider build is 14971; you have build 14393.

In fact, build 14393 is not an Insider build at all.  It is a regular Windows build.  Perhaps you're not familiar with the Insider program?  If not, read [here](https://insider.windows.com/) for more details, including information about how to join.  You can also download a VM with an Insider build (not the latest but relatively recent) [here](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) -- choose the "Win 10 Preview" one.  Note that Insider builds are pre-release builds of the entire Windows platform; they bring new features, but also more bugs than you would see in stable Windows releases.

This ticket was resolved in a recent Insider build.  The convention of this bugtracker and this community is that tickets are closed when an issue is resolved in an Insider build, as a signal that we can go test / play with it and confirm that it works as the Windows team expects.  The current stable release of Windows (known as the "Anniversary Edition") is too old to have the fix, and, because WSL is formally a Beta feature, fixes are only being pushed out via the Insider program and not via Windows Update.  These fixes (and there are lots and lots of them) will be released to regular stable Windows as part of the next major Windows release, sometime probably in the first half of next year.

So, in short:  If you're able to reproduce this in the latest Insider build, let us know.  If you can't or don't want to participate in the Insider program, be patient; it'll be a few months but you'll get the fix eventually.

**Comment by amrith** [comment]
Thanks Adam, no I’m not an insider and no, I was not looking to participate in the program. I’ll wait till a fix is available in the public builds. Can’t handle being on multiple insider programs at the same time; sqlserver-on-linux is enough for me right now ;)

 

Thanks for the detailed clarification(s),

 

-amrith

**Comment by ryanerwin** [comment]
This certainly isn't a perfect solution, but I've found that `httping` makes for a very nice replacement for almost everything that I use `ping` for and it works quite well under WSL.

I think there's even a httping package under Ubuntu `apt install httping` but you can just compile it from source.

`httping` even has a beautiful full screen mode (think like "top" for your network connection) and a color mode.

You can see my notes about [compiling it yourself under WSL](https://github.com/flok99/httping/issues/22) also on github.

**Comment by stephencnix** [comment]
@ryanerwin Problem is that httping is sending HTTP requests and not ICMP. This works well for pinging web-enabled servers/devices, but not for monitoring equipment/ISP connections that don't respond to HTTP requests.

**Comment by RyanBreaker** [comment]
Just wanted to say thanks for the work on this issue guys, I've been following this as I'd love to be able to use WSL for troubleshooting network issues but am not a fan of having to go the extra steps of opening it as an administrator each time I want to open it. I'm running the latest stable build of Windows 10, all updates and etc. installed, ICMP works when opened as admin but permission denied otherwise, same with traceroute; other network access works fine otherwise though.

**Comment by sunilmut** [comment]
@RyanBreaker - You are welcome and thanks for using WSL and providing feedback. Looks like you are using the Anniversary Update Windows 10 build (14393). Admin is no longer required for ping in the [Windows Insider Builds](https://blogs.windows.com/blog/tag/windows-insider-program/#VlmWmuwOR857jeEZ.97). But, these changes are not yet available in a general public release build. These changes will be part of the next general public release.

**Comment by thebigsmileXD** [comment]
@sunilmut I am in an insider update cycle, and i still am not able to use ping as non admin/as admin and neither to open/forward udp sockets. 
![image](https://cloud.githubusercontent.com/assets/8733998/21673209/e913d1bc-d327-11e6-8d8b-685b380703c8.png)

**Comment by aseering** [comment]
@thebigsmileXD -- What specific Insider build number are you running?  ("OS Build")

**Comment by sunilmut** [comment]
@thebigsmileXD - One more piece of information. Admin requirement for `ping` was lifted in the Insider build 14926, see [release notes](https://msdn.microsoft.com/en-us/commandline/wsl/release_notes#build-14926). Also, note that `sudo` is currently not anonymous to Windows admin.

**Comment by sunilmut** [comment]
@thebigsmileXD - Thanks for the info. 14393 is the Anniversary update build, and not an insider build. Please head to the Windows Insider blog to make sure that you are correctly signed up for the Insider update. 14986 is currently the latest Insider build.

**Comment by thebigsmileXD** [comment]
Weird that it didnt load it, thanks alot though!

**Comment by engjedi** [comment]
Hey folks, is there any limitation or blocking with respect to cmd bash output redirection?
I tried to run a simple "script" with ping and filter with grep and give a redirect of the output of the command to a file, and it goes blank.
One detail is that if I redirect a command that works "ex: ls -la > file" it redirects the content.
The contents of the script
#! / Bin / bash
For i in {1..254};do
Echo -n "[$ i]"
Ping 172.15.31. $ I | Grep "64 bytes"> ping
done
EOF
I would like to show a total with wc -l, as there are 23 alive, searching for redirection, I did not find any answer.

**Comment by aseering** [comment]
Hi @engjedi , thanks for posting!  This might actually be a better question for somewhere like StackOverflow, though:  No, there are no restrictions on piping.  But you appear to have some bugs in your script:  `>` doesn't append; it overwrites, you would need `>>`.  Also, Linux `ping` pings forever by default; you would need `ping -c 1`.  (Also, Linux `ping` only works if WSL is run as a Windows Administrator in the current stable build of Windows; this is fixed in current Insider builds.)  There are also issues with spacing and capitalization; I assume that's just a copy&paste issue, but in general it's helpful if you paste your exact script surrounded by triple backticks (\`\`\` code \`\`\`) so that it shows up as an unformatted block comment and folks can read it exactly.

In any case, most of those issues are general Linux shell-scripting concerns.  I'd encourage you to experiment with your script a bit more and ask around on a shell-scripting forum.  If you've tested your script on Linux and it works, and you've copied the script to WSL and it doesn't work, and you've narrowed it down to a particular part of the script (if it's a large script), and there's not an obvious existing ticket here covering the behavior, then at that point you've come to the right place; please go ahead and file a new ticket here and people will be happy to help.

**Comment by wartillo** [comment]
Hey,

I didn't read everything what people said about that thread, but i was facing that situation and to solve it, i just had to run bash as administrator. 

I wish that could help

**Comment by aseering** [comment]
Hey @wartillo -- thanks for the observation!  It sounds like you are not running an [Insider build](http://insider.windows.com)?  (Many people here are, though certainly many others aren't.)

**Comment by wartillo** [comment]
Hey Assering,

Forgot to specify version that i'm using. 
i'm using the Anniversary build.

Microsoft Windows [version 10.0.14393]

**Comment by karpulix** [comment]
if run Bash On Windows from administrator, command "ping" working correct.
(sorry for my english)

**Comment by subhammridha** [comment]
Yes. Same here. But I don't understand why?

**Comment by sunilmut** [comment]
@karpulix, @subhammridha - Which windows build are running? Please share output of the `ver` command from cmd. It seems like both of you are on Anniversay Update build (14393). In [Creators Update 15063](https://blogs.windows.com/windowsexperience/2017/04/11/how-to-get-the-windows-10-creators-update/#Oji9vxIYFbaV4GhJ.97), admin is no longer needed to run ping.

**Comment by neguetam** [comment]
just run " Bash ubuntu for windows " as administrator and that's it he can get permission for using socket

**Comment by subhammridha** [comment]
@sunilmut, yes I'm using anniversary update build of windows 10 and have not yet upgraded to creators update. I will though. 

Btw thanks for the info.

**Comment by karpulix** [comment]
@sunilmut  i use version 1607 (OS build 14393.1066). I can't use "ping" in bash if run bash without administrator permission. Maybe it depends from region updates. I'm from Russia. 

p.s. sorry for my english

**Comment by neguetam** [comment]
no problem mate hope it helps you 

good luck

**Comment by sunilmut** [comment]
@karpulix - No worries, thanks for using WSL. See if the [`Creators Update`](https://blogs.windows.com/windowsexperience/2017/04/11/how-to-get-the-windows-10-creators-update/#Oq31E9BiDi1Uy0Fr.97) is available to you or you can install it. It should solve your problems with ping and many other.

**Comment by karpulix** [comment]
@sunilmut Thank you! I'm update system and it problem resolved.

**Comment by brtmrl** [comment]
No need to update. Login as an administrator not as a guest.

**Comment by diveyez** [comment]
WSL has been nothing but problems.

**Comment by ryanerwin** [comment]
@diveyez 

ping has been working for several months, I think since the first creators update. 

WSL is still not perfect, but it's pretty impressive to me that compiled Linux binaries can use the Windows kernel in place of the Linux kernel and most of it works.

You can always run a VM if WSL doesn't deliver what you need. I'm eager for when WSL support is complete - no bugs at all. Personally I already find it less frustrating than trying to build Linux tools to run natively on Mac OS X, and it's getting better with each release.

**Comment by spacemishka** [comment]
I am trying to setup WSL but getting still errors on ping

`user@nb-999:~$ ping google.de
ping: unknown host google.de
user@nb-999:~$ ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
ping: sendmsg: Invalid argument
ping: sendmsg: Invalid argument`

resolv.conf:

`user@nb-999:~$ cat /etc/resolv.conf
nameserver 8.8.8.8
nameserver 8.8.4.4`

**Comment by aseering** [comment]
Hi @spacemishka -- what version/build of Windows are you running?  (Type `ver` at a Windows Command Prompt.)

This is a common issue with older releases of Windows 10, but a fix was released via Windows Update.

**Comment by Bravo13** [comment]
Same problem - no ping, nslookup etc. 
in bash
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
ping: sendmsg: Permission denied
ping: recvmsg: Permission denied
ping: recvmsg: Permission denied
ping: sendmsg: Permission denied
ping: recvmsg: Permission denied
ping: recvmsg: Permission denied

Microsoft Windows [Version 10.0.17134.48]

**Comment by mrmirnov** [comment]
Hi!

Still cannot run ping even if run-as admin mode until sudo it:
Microsoft Windows [Version 10.0.16299.726]
msmirnov@smirnov:/mnt/c/Windows/System32$ ping www.ru
ping: socket: Operation not permitted
msmirnov@smirnov:/mnt/c/Windows/System32$ ping www.google.com
ping: socket: Operation not permitted

**Comment by sxpistols** [comment]
> ```shell
> sudo chmod u+s `which ping`
> ```

thank you, it worked for me

**Comment by Dibbyo456** [comment]
> for anyone still running into this issue, the below command fixed it for me:
> 
> ```shell
> sudo chmod u+s `which ping`
> ```

Installed Ubuntu 20.04 on WSL 2, ran into the same issue. But that solution worked. 👍

**Comment by fuomag9** [comment]
> 
> 
> for anyone still running into this issue, the below command fixed it for me:
> 
> ```shell
> sudo chmod u+s `which ping`
> ```

Brand new windows 2004 kali wsl2 install, this fixed it for me as well, thanks!

**Comment by mspncp** [comment]
FWIW:  according to the manual page,  it  is sufficient to set the CAP_NET_RAW capability. The following command worked for me:

    sudo setcap cap_net_raw+ep `which ping`

Whether this method is safer than 'set-uid root' or not, I can't tell.

**Comment by toor11** [comment]
That is a permission issue:
Before : 
user@localhost:~$ for n in {1..5}; do ping 1.1.1.1; done
ping: socket: Operation not permitted
ping: socket: Operation not permitted
ping: socket: Operation not permitted
ping: socket: Operation not permitted
ping: socket: Operation not permitted

AFTER:
user@localhost:~$ sudo setcap cap_net_raw+p /bin/ping
[sudo] password for user:
user@localhost:~$ ping 1.1.1.1
PING 1.1.1.1 (1.1.1.1) 56(84) bytes of data.
64 bytes from 1.1.1.1: icmp_seq=1 ttl=50 time=46.2 ms
64 bytes from 1.1.1.1: icmp_seq=2 ttl=50 time=45.0 ms
